### ISSUE TITLE
ImportError: DLL load failed while importing pybind: A dynamic link library (DLL) initialization routine failed.

### ISSUE THREAD

**Issue by dorovl** [isl-org/Open3D#4468]
```
D:\>pip install --user --pre https://storage.googleapis.com/open3d-releases-master/python-wheels/open3d-0.14.1+da463fb-cp39-cp39-win_amd64.whl

D:\>python -c "import open3d as o3d; print(o3d)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\vlada\AppData\Roaming\Python\Python39\site-packages\open3d\__init__.py", line 90, in <module>
    from open3d.cpu.pybind import (camera, geometry, io, pipelines, utility, t)
ImportError: DLL load failed while importing pybind: A dynamic link library (DLL) initialization routine failed.

D:\>pip install open3d==0.14.1

D:\>python -c "import open3d as o3d; print(o3d)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\vlada\AppData\Roaming\Python\Python39\site-packages\open3d\__init__.py", line 90, in <module>
    from open3d.cpu.pybind import (camera, geometry, io, pipelines, utility, t)
ImportError: DLL load failed while importing pybind: A dynamic link library (DLL) initialization routine failed.
```

- Operating system: Windows 10 64-bit
- Python version: Python 3.9.9
- Open3D version: 0.14.1
- Is this remote workstation?: no
- How did you install Open3D?: pip

**Comment by dorovl** [comment]
If I run the following program:
```
import dlltracer
import sys

with dlltracer.Trace(out=sys.stdout):
    import open3d as o3d
    print(o3d)
```

I obtain:

```
LoadLibrary \Device\HarddiskVolume2\Windows\System32\kernel.appcore.dll
LoadLibrary \Device\HarddiskVolume2\Program Files\Python39\DLLs\_ctypes.pyd
LoadLibrary \Device\HarddiskVolume2\Windows\System32\ole32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\combase.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\gdi32.dll
LoadLibrary \Device\HarddiskVolume2\Program Files\Python39\DLLs\libffi-7.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\win32u.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\gdi32full.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\msvcp_win.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\user32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\oleaut32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\imm32.dll
LoadLibrary \Device\HarddiskVolume2\Program Files\Python39\DLLs\_bz2.pyd
LoadLibrary \Device\HarddiskVolume2\Program Files\Python39\DLLs\_lzma.pyd
LoadLibrary \Device\HarddiskVolume2\Users\vlada\AppData\Roaming\Python\Python39\site-packages\open3d\cpu\pybind.cp39-win_amd64.pyd
LoadLibrary \Device\HarddiskVolume2\Windows\System32\shell32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\comdlg32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\SHCore.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\secur32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\shlwapi.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\winmm.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\opengl32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\setupapi.dll
LoadLibrary \Device\HarddiskVolume2\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\comctl32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\glu32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\vcomp140.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\mf.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\mfplat.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\mfreadwrite.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\winusb.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\sspicli.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\mfcore.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\crypt32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\powrprof.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\ksuser.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\RTWorkQ.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\umpdc.dll
Failed \Device\HarddiskVolume2\Windows\System32\opengl32.dll
Failed \Device\HarddiskVolume2\Windows\System32\glu32.dll
Failed \Device\HarddiskVolume2\Windows\System32\secur32.dll
Failed \Device\HarddiskVolume2\Windows\System32\winmm.dll
Failed \Device\HarddiskVolume2\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\comctl32.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\windows.storage.dll
LoadLibrary \Device\HarddiskVolume2\Windows\System32\wldp.dll
Failed \Device\HarddiskVolume2\Windows\System32\wldp.dll
Failed \Device\HarddiskVolume2\Windows\System32\windows.storage.dll
Failed \Device\HarddiskVolume2\Windows\System32\shell32.dll
Failed \Device\HarddiskVolume2\Windows\System32\comdlg32.dll
Failed \Device\HarddiskVolume2\Windows\System32\vcomp140.dll
Failed \Device\HarddiskVolume2\Windows\System32\shlwapi.dll
Failed \Device\HarddiskVolume2\Windows\System32\setupapi.dll
Failed \Device\HarddiskVolume2\Windows\System32\mfreadwrite.dll
Failed \Device\HarddiskVolume2\Windows\System32\winusb.dll
Failed \Device\HarddiskVolume2\Windows\System32\sspicli.dll
Failed \Device\HarddiskVolume2\Users\vlada\AppData\Roaming\Python\Python39\site-packages\open3d\cpu\pybind.cp39-win_amd64.pyd
Traceback (most recent call last):
  File "D:\depthai-experiments\pcl-projection-rgb\test.py", line 5, in <module>
    import open3d as o3d
  File "C:\Users\vlada\AppData\Roaming\Python\Python39\site-packages\open3d\__init__.py", line 90, in <module>
    from open3d.cpu.pybind import (camera, geometry, io, pipelines, utility, t)
ImportError: DLL load failed while importing pybind: A dynamic link library (DLL) initialization routine failed.
```

**Comment by dorovl** [comment]
The similar installation on a newer computer works. I suspect a problem with the graphical card support of opengl32.dll : this PC has a rather old graphics card (NVIDIA GeForce GTS 360M). The GPU driver is up to date.

**Comment by ssheorey** [comment]
@dorovl what's the driver version and highest supported OpenGL version? Open3D needs OpenGL 4.1.

**Comment by jackjansen** [comment]
I'm running into this same issue. I'm using Open3D on Windows, running in a Parallels VM.

Parallels supports only OpenGL 3.3.

Is there a way to tell Open3D not to use the GPU, or to use DirectX, or something else?

I'm not really interested in performance: this is my CI/CD machine, the software will run on capable machines in production but it would be a bit of a bother to turn off all tests in the CI/CD...

**Comment by ssheorey** [comment]
We have recently added CPU rendering capability to Open3D for use cases such as yours. It's not ready for production yet and is only tested on Linux (PR #4880). The basic idea is to preload the Mesa OpenGL DLL that supports CPU rendering before starting your application. This is not recommended in [Windows](https://docs.microsoft.com/en-US/windows/win32/dlls/secure-boot-and-appinit-dlls), I believe. We are still working on this...

**Comment by jackjansen** [comment]
@ssheorey I understand that it isn't recommended. It seems it doesn't work:-)

I've downloaded the only MESA for Windows I could find, from https://fdossena.com/?p=mesa/index.frag, but open3d still won't import:

```
>>> import ctypes
>>> ctypes.cdll.LoadLibrary("c:/mytemp/opengl32.dll")
<CDLL 'c:\mytemp\opengl32.dll', handle 7ffdd8ec0000 at 0x19659689fd0>
>>> import open3d
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python39\lib\site-packages\open3d\__init__.py", line 90, in <module>
    from open3d.cpu.pybind import (camera, data, geometry, io, pipelines,
ImportError: DLL load failed while importing pybind: A dynamic link library (DLL) initialization routine failed.
```

I wouldn't be surprised if the problem is that (since Python 3.7) the Windows DLL search algorithm has become very convoluted (with good reasons from a security point of view). In another project (actually the project for which I'm now trying to get the CI/CD working) I've had to revert to manually adding all search directories with `os.add_dll_directory()`. So maybe the preloaded `opengl32.dll` is ignored.

I will try with `dlltracer`.

**Comment by ssheorey** [comment]
Thanks for trying it. Let me know if you are able to get it to work and we will be happy to include it in master. We are still ironing out some problems on Linux docker as well.

**Comment by jackjansen** [comment]
Unfortunately the problem is more difficult. The preloaded `opengl32.dll` is picked up, but other things fail:

```

>>> import ctypes
>>> import dlltracer
>>> import sys
>>> with dlltracer.Trace(out=sys.stdout):
...     ctypes.cdll.LoadLibrary("c:/mytemp/opengl32.dll")
...
<CDLL 'c:\mytemp\opengl32.dll', handle 7ffdd8ec0000 at 0x232da177eb0>
LoadLibrary \Device\HarddiskVolume4\Windows\System32\kernel.appcore.dll
LoadLibrary \Device\HarddiskVolume4\mytemp\opengl32.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\shell32.dll
>>> with dlltracer.Trace(out=sys.stdout):
...     import open3d
...
LoadLibrary \Device\HarddiskVolume4\Python39\DLLs\_bz2.pyd
LoadLibrary \Device\HarddiskVolume4\Python39\DLLs\_lzma.pyd
LoadLibrary \Device\HarddiskVolume4\Python39\Lib\site-packages\open3d\cpu\pybind.cp39-win_amd64.pyd
LoadLibrary \Device\HarddiskVolume4\Windows\System32\comdlg32.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\SHCore.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\winmm.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\secur32.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\shlwapi.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\crypt32.dll
LoadLibrary \Device\HarddiskVolume4\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\comctl32.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\setupapi.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\mf.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\mfplat.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\mfreadwrite.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\winusb.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\vcomp140.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\sspicli.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\mfcore.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\powrprof.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\ksuser.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\RTWorkQ.dll
LoadLibrary \Device\HarddiskVolume4\Windows\System32\umpdc.dll
Failed \Device\HarddiskVolume4\Windows\System32\secur32.dll
Failed \Device\HarddiskVolume4\Windows\System32\winmm.dll
Failed \Device\HarddiskVolume4\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\comctl32.dll
Failed \Device\HarddiskVolume4\Windows\System32\comdlg32.dll
Failed \Device\HarddiskVolume4\Windows\System32\vcomp140.dll
Failed \Device\HarddiskVolume4\Windows\System32\shlwapi.dll
Failed \Device\HarddiskVolume4\Windows\System32\setupapi.dll
Failed \Device\HarddiskVolume4\Windows\System32\mfreadwrite.dll
Failed \Device\HarddiskVolume4\Windows\System32\winusb.dll
Failed \Device\HarddiskVolume4\Windows\System32\sspicli.dll
Failed \Device\HarddiskVolume4\Python39\Lib\site-packages\open3d\cpu\pybind.cp39-win_amd64.pyd
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "C:\Python39\lib\site-packages\open3d\__init__.py", line 90, in <module>
    from open3d.cpu.pybind import (camera, data, geometry, io, pipelines,
ImportError: DLL load failed while importing pybind: A dynamic link library (DLL) initialization routine failed.
>>>
```

I've reached the end of my Windows hacking skills. Let's hope someone else wanders by here who understands more about it.

**Comment by Suisse00** [comment]
Hi, I have a similar problem as this guy (though, I tried to use python 3.9.5).
Tried to just load opengl and got similar error (thought address was different, could be because of a different version or because my bug isn't related at all).

I tried open3d version (from pip in a venv):

- 15.1 (latest for windows, 15.2 is only linux)
- 14.1 (the only other version available for windows x64 on python 3.9)

I also tried to compile the master (998464c4b0311aab11d8669726cca5c6483c6e6) to try to get some debugging symbol.

In every case still got the error.

I tried to plug Visual Studio debugger into the mix.

on 15.1:
```
Exception thrown at 0x00007FFD0828A9A0 (pybind.cp39-win_amd64.pyd) in python.exe: 0xC000001D: Illegal Instruction.
	pybind.cp39-win_amd64.pyd!00007ffd0828a9a0()	Unknown
 00007FFD0828A9A0  vpxor       xmm0,xmm0,xmm0  
```

on the compiled master branch:
```
Exception thrown at 0x00007FFD061CF222 (pybind.cp39-win_amd64.pyd) in python.exe: 0xC000001D: Illegal Instruction.
	pybind.cp39-win_amd64.pyd!embree::PrecomputedBezierBasis::PrecomputedBezierBasis(int dj) Line 9	C++
00007FFD061CF222  vmovaps     xmmword ptr [rsp+40h],xmm6  
```
(They were `vpxor` just after as well)
(Which is likely to match the same exact issue as https://github.com/isl-org/Open3D/issues/5435 but hidden on Windows)

After Googling, `vpxor` is part of the AVX2 instructions and `vmovaps` from AVX-256 (AVX?).

I checked the datasheet of my old potato i7 920, nothing about AVX.  [CPU-Z](https://www.cpuid.com/softwares/cpu-z.html) confirmed me the same thing.

So now, I have to figure out how to compile it without those optimization to see if I can still _use it_.

**Comment by BastianGit** [comment]
> Unfortunately the problem is more difficult. The preloaded `opengl32.dll` is picked up, but other things fail:
> 
> ```
> 
> >>> import ctypes
> >>> import dlltracer
> >>> import sys
> >>> with dlltracer.Trace(out=sys.stdout):
> ...     ctypes.cdll.LoadLibrary("c:/mytemp/opengl32.dll")
> ...
> <CDLL 'c:\mytemp\opengl32.dll', handle 7ffdd8ec0000 at 0x232da177eb0>
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\kernel.appcore.dll
> LoadLibrary \Device\HarddiskVolume4\mytemp\opengl32.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\shell32.dll
> >>> with dlltracer.Trace(out=sys.stdout):
> ...     import open3d
> ...
> LoadLibrary \Device\HarddiskVolume4\Python39\DLLs\_bz2.pyd
> LoadLibrary \Device\HarddiskVolume4\Python39\DLLs\_lzma.pyd
> LoadLibrary \Device\HarddiskVolume4\Python39\Lib\site-packages\open3d\cpu\pybind.cp39-win_amd64.pyd
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\comdlg32.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\SHCore.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\winmm.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\secur32.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\shlwapi.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\crypt32.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\comctl32.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\setupapi.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\mf.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\mfplat.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\mfreadwrite.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\winusb.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\vcomp140.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\sspicli.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\mfcore.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\powrprof.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\ksuser.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\RTWorkQ.dll
> LoadLibrary \Device\HarddiskVolume4\Windows\System32\umpdc.dll
> Failed \Device\HarddiskVolume4\Windows\System32\secur32.dll
> Failed \Device\HarddiskVolume4\Windows\System32\winmm.dll
> Failed \Device\HarddiskVolume4\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\comctl32.dll
> Failed \Device\HarddiskVolume4\Windows\System32\comdlg32.dll
> Failed \Device\HarddiskVolume4\Windows\System32\vcomp140.dll
> Failed \Device\HarddiskVolume4\Windows\System32\shlwapi.dll
> Failed \Device\HarddiskVolume4\Windows\System32\setupapi.dll
> Failed \Device\HarddiskVolume4\Windows\System32\mfreadwrite.dll
> Failed \Device\HarddiskVolume4\Windows\System32\winusb.dll
> Failed \Device\HarddiskVolume4\Windows\System32\sspicli.dll
> Failed \Device\HarddiskVolume4\Python39\Lib\site-packages\open3d\cpu\pybind.cp39-win_amd64.pyd
> Traceback (most recent call last):
>   File "<stdin>", line 2, in <module>
>   File "C:\Python39\lib\site-packages\open3d\__init__.py", line 90, in <module>
>     from open3d.cpu.pybind import (camera, data, geometry, io, pipelines,
> ImportError: DLL load failed while importing pybind: A dynamic link library (DLL) initialization routine failed.
> >>>
> ```
> 
> I've reached the end of my Windows hacking skills. Let's hope someone else wanders by here who understands more about it.

Hi,

I had a similar issue during import of open3d version 0.16 and 0.15.1 with windows 11 and python 9.13. I managed to solve the problem by installing CUDA version 11.5 and setting the paths correctly in the system environment variables (cuda installation should be similar to: c>program files> nvidia gpu computing... > cuda > v11.5). Uninstall open3d and re-install.

I hope this helps!
Regards.

**Comment by JonatanPlantey** [comment]
After several tries, the only thing that worked for me was install CUDA (and visual studio before, because demanded by CUDA). I had forgotten that wasn't installed since my reinstallation of windows operating system...

**Comment by rvenky125** [comment]
is there any fixes?

**Comment by nigels-com** [comment]
Would be nice to have a way to build Open3D _without_ an OpenGL dependency.  For batch processing scenarios, we don't use the OpenGL functionality at all.

**Comment by hyrodium** [comment]
The error was fixed by updating graphics drivers with Intel DSA on my environment.

https://www.intel.com/content/www/us/en/support/detect.html
